activation/activation_multi_test.go

```diff
@@ -0,0 +1,483 @@
+
+	for _, sig := range tab.signers {
+		tab.mnipost.EXPECT().ResetState(sig.NodeID()).Return(nil)
+	}
+	require.NoError(t, tab.StopSmeshing(true))
+}
+
+func Test_Builder_Multi_RestartSmeshing(t *testing.T) {
+	getBuilder := func(t *testing.T) *Builder {
+		tab := newTestBuilder(t, 5)
+
+		for _, sig := range tab.signers {
+			tab.mnipost.EXPECT().Proof(gomock.Any(), sig.NodeID(), shared.ZeroChallenge).AnyTimes().DoAndReturn(
+				func(ctx context.Context, _ types.NodeID, _ []byte) (*types.Post, *types.PostInfo, error) {
+					<-ctx.Done()
+					return nil, nil, ctx.Err()
+				})
+		}
+
+		ch := make(chan struct{})
+		close(ch)
+		tab.mclock.EXPECT().AwaitLayer(gomock.Any()).Return(ch).AnyTimes()
+		tab.mclock.EXPECT().CurrentLayer().Return(types.LayerID(0)).AnyTimes()
+		return tab.Builder
+	}
+
+	t.Run("Single threaded", func(t *testing.T) {
+		builder := getBuilder(t)
+		for i := 0; i < 50; i++ {
+			require.NoError(t, builder.StartSmeshing(types.Address{}))
+			require.True(t, builder.Smeshing())
+			require.NoError(t, builder.StopSmeshing(false))
+			require.False(t, builder.Smeshing())
+		}
+	})
+
+	t.Run("Multi threaded", func(t *testing.T) {
+		// Meant to be run with -race to detect races.
+		// It cannot check `builder.Smeshing()` as Start/Stop is happening from many goroutines simultaneously.
+		// Both Start and Stop can fail as it is not known if builder is smeshing or not.
+		builder := getBuilder(t)
+		var eg errgroup.Group
+		for worker := 0; worker < 10; worker += 1 {
+			eg.Go(func() error {
+				for i := 0; i < 50; i++ {
+					builder.StartSmeshing(types.Address{})
+					builder.StopSmeshing(false)
+				}
+				return nil
+			})
+		}
+		require.NoError(t, eg.Wait())
+	})
+}
+
+func Test_Builder_Multi_StopSmeshing_Delete(t *testing.T) {
+	tab := newTestBuilder(t, 5)
+
+	atx := types.RandomATXID()
+	refChallenge := &types.NIPostChallenge{
+		PublishEpoch:  postGenesisEpoch + 2,
+		CommitmentATX: &atx,
+	}
+
+	currLayer := (postGenesisEpoch + 1).FirstLayer()
+	tab.mclock.EXPECT().AwaitLayer(gomock.Any()).Return(make(chan struct{})).AnyTimes()
+	tab.mclock.EXPECT().CurrentLayer().Return(currLayer).AnyTimes()
+	tab.mclock.EXPECT().LayerToTime(gomock.Any()).DoAndReturn(
+		func(got types.LayerID) time.Time {
+			// time.Now() ~= currentLayer
+			genesis := time.Now().Add(-time.Duration(currLayer) * layerDuration)
+			return genesis.Add(layerDuration * time.Duration(got))
+		}).AnyTimes()
+
+	for _, sig := range tab.signers {
+		tab.mnipost.EXPECT().Proof(gomock.Any(), sig.NodeID(), shared.ZeroChallenge).DoAndReturn(
+			func(ctx context.Context, _ types.NodeID, _ []byte) (*types.Post, *types.PostInfo, error) {
+				<-ctx.Done()
+				return nil, nil, ctx.Err()
+			})
+
+		// add challenge to DB
+		require.NoError(t, nipost.AddChallenge(tab.localDb, sig.NodeID(), refChallenge))
+	}
+
+	require.NoError(t, tab.StartSmeshing(types.Address{}))
+	require.NoError(t, tab.StopSmeshing(false))
```

Comment:
Would it make sense to persist the gomock Controller and assert that all expectations are satisfied so far? like:
```go
ctrl := gomock.NewController(t)
tab := newTestBuilder(t, ctrl, 5)
...
require.NoError(t, tab.StopSmeshing(false))
require.True(t, ctrl.Satisfied())
```